lib/evaluate/rte.cc

```diff
@@ -0,0 +1,166 @@
+// limitations under the License.
+
+// This file defines the runtime libraries for the target as well as a default
+// set of host rte functions that can be used for folding.
+// The default HostRte is built with <cmath> and <complex> functions
+// that are guaranteed to exist from the C++ standard.
+
+#include "rte.h"
+#include <cerrno>
+#include <cfenv>
+
+namespace Fortran::evaluate::rte {
+
+using namespace Fortran::parser::literals;
+// TODO mapping to <cmath> function to be tested.<cmath> func takes
+// real arg for n
+template<typename HostT> static HostT Bessel_jn(std::int64_t n, HostT x) {
+  return std::cyl_bessel_j(static_cast<HostT>(n), x);
+}
+template<typename HostT> static HostT Bessel_yn(std::int64_t n, HostT x) {
+  return std::cyl_neumann(static_cast<HostT>(n), x);
+}
+
+template<typename HostT> void AddLibmRealHostProcedure(HostRte &hostRte) {
+  using F = FuncPointer<HostT, HostT>;
+  using F2 = FuncPointer<HostT, HostT, HostT>;
+  HostRteProcedureSymbol libmSymbols[]{{"acos", F{std::acos}, true},
+      {"acosh", F{std::acosh}, true}, {"asin", F{std::asin}, true},
+      {"asinh", F{std::asinh}, true}, {"atan", F{std::atan}, true},
+      {"atan", F2{std::atan2}, true}, {"atanh", F{std::atanh}, true},
+      {"bessel_jn", &Bessel_jn<HostT>, true},
+      {"bessel_yn", &Bessel_yn<HostT>, true}, {"cos", F{std::cos}, true},
+      {"cosh", F{std::cosh}, true}, {"sin", F{std::sin}, true},
+      {"sinh", F{std::sinh}, true}, {"sqrt", F{std::sqrt}, true},
+      {"tan", F{std::tan}, true}, {"tanh", F{std::tanh}, true}};
+
+  for (auto sym : libmSymbols) {
+    hostRte.AddProcedure(std::move(sym));
+  }
+}
+
+template<typename HostT> void AddLibmComplexHostProcedure(HostRte &hostRte) {
+  using F = FuncPointer<std::complex<HostT>, const std::complex<HostT> &>;
+  HostRteProcedureSymbol libmSymbols[]{{"acos", F{std::acos}, true},
+      {"acosh", F{std::acosh}, true}, {"asin", F{std::asin}, true},
+      {"asinh", F{std::asinh}, true}, {"atan", F{std::atan}, true},
+      {"atanh", F{std::atanh}, true}, {"cos", F{std::cos}, true},
+      {"cosh", F{std::cosh}, true}, {"sin", F{std::sin}, true},
+      {"sinh", F{std::sinh}, true}, {"sqrt", F{std::sqrt}, true},
+      {"tan", F{std::tan}, true}, {"tanh", F{std::tanh}, true}};
+
+  for (auto sym : libmSymbols) {
+    hostRte.AddProcedure(std::move(sym));
+  }
+}
+
+void HostRte::DefaultInit() {
+  AddLibmRealHostProcedure<float>(*this);
+  AddLibmRealHostProcedure<double>(*this);
+  AddLibmRealHostProcedure<long double>(*this);
+
+  AddLibmComplexHostProcedure<float>(*this);
+  AddLibmComplexHostProcedure<double>(*this);
+  AddLibmComplexHostProcedure<long double>(*this);
+}
+
+void HostFloatingPointEnvironment::SetUpHostFloatingPointEnvironment(
+    FoldingContext &context) {
+  errno = 0;
+  if (feholdexcept(&originalFenv_) != 0) {
+    context.messages().Say(
+        "Folding with host runtime: feholdexcept() failed: %s"_en_US,
+        std::strerror(errno));
```

Comment:
What can the user do when they get this error?

Reply:
To be honest, I did not find why this function could failed exactly apart that from its definition it can. So I am not entirely sure what are the consequences of a failure and what should happen. I have no idea what the user should do appart from compiling again and reporting if the issue persists.
Would you make it a fatal error to be safe?

Reply:
I think I would treat it as an internal error until we find a case where it can actually happen.

Reply:
Done in commit f776c8d. I changed all errors to internal error.
I only left a warning when the rounding mode cannot be set as expected because it does not exists. What happens then is well defined, the user just needs to know that the results might not be bit to bit comparable to what he expects.